kernel/src/table_properties.rs

```diff
@@ -0,0 +1,373 @@
+    /// size increases.
+    #[serde(rename = "delta.logRetentionDuration")]
+    #[serde(deserialize_with = "deserialize_interval")]
+    pub log_retention_duration: Option<Duration>,
+
+    /// TODO docs
+    #[serde(rename = "delta.enableExpiredLogCleanup")]
+    #[serde(deserialize_with = "deserialize_bool")]
+    pub enable_expired_log_cleanup: Option<bool>,
+
+    /// true for Delta to generate a random prefix for a file path instead of partition information.
+    ///
+    /// For example, this may improve Amazon S3 performance when Delta Lake needs to send very high
+    /// volumes of Amazon S3 calls to better partition across S3 servers.
+    #[serde(rename = "delta.randomizeFilePrefixes")]
+    #[serde(deserialize_with = "deserialize_bool")]
+    pub randomize_file_prefixes: Option<bool>,
+
+    /// When delta.randomizeFilePrefixes is set to true, the number of characters that Delta
+    /// generates for random prefixes.
+    #[serde(rename = "delta.randomPrefixLength")]
+    #[serde(deserialize_with = "deserialize_pos_int")]
+    pub random_prefix_length: Option<u64>,
+
+    /// The shortest duration within which new snapshots will retain transaction identifiers (for
+    /// example, SetTransactions). When a new snapshot sees a transaction identifier older than or
+    /// equal to the duration specified by this property, the snapshot considers it expired and
+    /// ignores it. The SetTransaction identifier is used when making the writes idempotent.
+    #[serde(rename = "delta.setTransactionRetentionDuration")]
+    #[serde(deserialize_with = "deserialize_interval")]
+    pub set_transaction_retention_duration: Option<Duration>,
+
+    /// The target file size in bytes or higher units for file tuning. For example, 104857600
+    /// (bytes) or 100mb.
+    #[serde(rename = "delta.targetFileSize")]
+    #[serde(deserialize_with = "deserialize_pos_int")]
+    pub target_file_size: Option<u64>,
+
+    /// The target file size in bytes or higher units for file tuning. For example, 104857600
+    /// (bytes) or 100mb.
+    #[serde(rename = "delta.tuneFileSizesForRewrites")]
+    #[serde(deserialize_with = "deserialize_bool")]
+    pub tune_file_sizes_for_rewrites: Option<bool>,
+
+    /// 'classic' for classic Delta Lake checkpoints. 'v2' for v2 checkpoints.
+    #[serde(rename = "delta.checkpointPolicy")]
+    #[serde(deserialize_with = "deserialize_option")]
+    pub checkpoint_policy: Option<CheckpointPolicy>,
+
+    /// whether to enable row tracking during writes.
+    #[serde(rename = "delta.enableRowTracking")]
+    #[serde(deserialize_with = "deserialize_bool")]
+    pub enable_row_tracking: Option<bool>,
+
+    /// any unrecognized properties are passed through and ignored by the parser
+    #[serde(flatten)]
+    pub unknown_properties: HashMap<String, String>,
+}
+
+impl TableProperties {
+    pub(crate) fn new(config_map: HashMap<String, String>) -> DeltaResult<Self> {
```

Comment:
You can make a `MapDeserializer` from an iterator of pairs of things that are [`serde::de::IntoDeserializer`](https://docs.rs/serde/latest/serde/de/trait.IntoDeserializer.html). `&'a str` is. The issue you were running into is that `&String` _isn't_, but if you map over and `as_str()` them it'll work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e8cb7975f1d98cec1f4a2452da5fba8c

So this could probably be:
```suggestion
    pub(crate) fn new(items: I ) -> DeltaResult<Self> 
    where
        I: Iterator<Item = (&'a str, &'a str)>
 {
```

It'd be nice to take the same signature as t`MapDeserializer::new`, but we can't because that refers to a type that's private in serde (which seems bad to me but 🤷🏽)

Reply:
yup got it with some minor tweaking! thank you!!